### PR TITLE
Addon Vitest: Add experimental vitest integration

### LINKED ISSUE #28388: [Tracking] Vitest integration
## RFC
https://github.com/storybookjs/storybook/discussions/28386

## Milestones

The Vitest integration effort is split into multiple types of work. It involves a Storybook plugin for Vitest, which transforms stories into tests (see RFC); Improvements in the core layer of Storybook; Some updates to the interactions addon and a creation of a new Storybook addon: `@storybook/addon-vitest`.

### Milestone 1 - Add filtering logic and improved UI to Storybook (extracted from VTA)

```[tasklist]
- [ ] https://github.com/storybookjs/storybook/issues/28682
- [ ] https://github.com/storybookjs/storybook/issues/28679
- [ ] https://github.com/storybookjs/storybook/issues/28680
- [ ] https://github.com/storybookjs/storybook/issues/28681
- [ ] https://github.com/storybookjs/storybook/issues/28683
```

### Milestone 2 - Portable stories and core improvements

```[tasklist]
- [ ] https://github.com/storybookjs/storybook/issues/28685
- [ ] https://github.com/storybookjs/storybook/pull/28764
- [ ] https://github.com/storybookjs/storybook/pull/28864
- [ ] https://github.com/storybookjs/storybook/pull/28829
- [ ] https://github.com/storybookjs/storybook/pull/28865
- [ ] https://github.com/storybookjs/storybook/pull/28825
- [x] metrics: https://github.com/storybookjs/storybook/pull/26764 [yann, michael]
```

### Milestone 3 - Experimental Vitest plugin - part 1
```[tasklist]
- [x] https://github.com/storybookjs/vitest-plugin/pull/1
- [x] https://github.com/storybookjs/vitest-plugin/pull/2
- [x] https://github.com/storybookjs/vite-plugin-storybook-nextjs/pull/3
- [ ] https://github.com/storybookjs/storybook/issues/28687
- [ ] https://github.com/storybookjs/storybook/issues/28707
- [ ] https://github.com/storybookjs/storybook/issues/28684
- [ ] https://github.com/storybookjs/storybook/issues/28744
- [ ] https://github.com/storybookjs/storybook/issues/28775
- [ ] https://github.com/storybookjs/storybook/issues/28777
- [ ] https://github.com/storybookjs/storybook/issues/28852
- [x] https://github.com/yannbf/mealdrop/pull/30
- [x] https://github.com/chromaui/chromatic/pull/8864
- [x] https://github.com/storybookjs/storybook-rsc-demo/pull/15
- [x] Use new `run` portable stories format (testingLibraryRender) [kasper, yann]
- [x] Write unit tests [yann]
- [x] Bug: Investigate double registration of plugin [yann]
- [x] Add unit testing/E2E testing/QA [kasper, yann]
- [x] metrics: make sure it's part of telemetry
```

### Milestone 4 - Experimental Vitest plugin - part 2
```[tasklist]
- [ ] https://github.com/storybookjs/storybook/issues/28807
- [ ] https://github.com/storybookjs/storybook/issues/28814
- [ ] https://github.com/storybookjs/storybook/issues/28889
- [ ] https://github.com/storybookjs/storybook/issues/28901
- [ ] https://github.com/storybookjs/storybook/issues/28903
- [ ] https://github.com/storybookjs/storybook/issues/28890
- [ ] https://github.com/storybookjs/storybook/issues/28904
- [ ] https://github.com/storybookjs/storybook/issues/28953
- [ ] **nice to have:** Research on isolation mode (workspace config)
- [ ] https://github.com/storybookjs/storybook/issues/28954
- [ ] https://github.com/storybookjs/storybook/issues/28930
```

### Milestone 5 - Documentation and content creation
```[tasklist]
- [ ] https://github.com/storybookjs/storybook/issues/28893
- [x] ViteConf: Prepare presentation [yann]
```

### Milestone 6 - Extended portable stories support
```[tasklist]
- [ ] https://github.com/storybookjs/storybook/issues/28895
- [ ] https://github.com/storybookjs/storybook/issues/28891
- [ ] https://github.com/storybookjs/storybook/issues/28892
- [ ] https://github.com/storybookjs/storybook/issues/28898
- [ ] https://github.com/storybookjs/storybook/issues/28897
- [ ] https://github.com/storybookjs/storybook/issues/28896
```

### Unplanned work

The following tasks were discussed to be out of scope, given the probable lack of time to work on them:

```[tasklist]
- [ ] https://github.com/storybookjs/storybook/issues/28688
- [ ] https://github.com/storybookjs/storybook/issues/28738
- [ ] https://github.com/storybookjs/storybook/issues/28686
- [ ] **Addon-interactions:** Improve error handling to allow non-playfn error logs [gert]
- [ ] **Vitest addon:** Panel with Vitest CLI raw logs [gert]
- [ ] **Vitest addon:** Pipe CLI output of Vitest onto Storybook's CLI
- [ ] **Vitest addon:** Implement Vitest orchestration layer [kasper, yann]
- [ ] **nice to have:** Move server connection checks to Storybook core [gert] https://github.com/storybookjs/storybook/issues/28258
- [ ] **nice to have:** Move top bar button (run tests) logic from VTA to Storybook [gert]
- [ ] **nice to have:** Remove/deprecate experimental topBar APIs in Storybook core [gert]
- [ ] **nice to have:** UI: Implement Superbar component
- [ ] **nice to have:** Implement afterEach as kind of a replacement for useEffect that is more compatible with test runner, as it can be await-ed. [kasper]
- [ ] **nice to have:** Audit all of our core addons and use the afterEach mechanism if applicable (and make PS compatible) [yann, kasper]
- [ ] **nice to have:** Implement a11y reporting mechanism [kasper]
- [ ] **nice to have:** Test package: Experiment with (inline) snapshots [kasper]
- [ ] Spike "always on watch mode" by running tests in watch mode via the CLI + watching for Junit files
```

### PR DESCRIPTION
**Pull request by kasperpeulen**
Closes #28388

## What I did

This is the umbrella PR that will cover most of the work related to the [Vitest integration project](https://github.com/storybookjs/storybook/issues/28388).

It includes a new Storybook addon called `@storybook/experimental-addon-vitest`, which currenly only includes a Vitest plugin. Later, it will include more functionality.

Additionally:

[Core: Enable RSC support when parameters.react is set to true #28865](https://github.com/storybookjs/storybook/pull/28865)

Now RSC can be enabled via `parameters.react.rsc: true`, as the flag is always on by default.

[UI: Replace experimental_SIDEBAR_BOTTOM API with a status filter UI #28693](https://github.com/storybookjs/storybook/pull/28693)

The bottom sidebar filter, which used to live in VTA, is now part of Storybook core. There should be no changes when using VTA with Storybook.

[UI: Update sidebar status indicators and tooltips #28739](https://github.com/storybookjs/storybook/pull/28739)

The status UI had a revamp with improved icons and styling.

[Portable stories: Use internal Storybook renderers mechanism #28766](https://github.com/storybookjs/storybook/pull/28766)

Portable stories now does not depend on testing-library.

### Testing

This is extensively tested in unit tests and sandbox stories tests. If you want to test it yourself, there are a couple ways:
1 - Create any vite-based sandbox. It will already have the plugin set-up and you can just run `yarn vitest` to run tests in.
2 - On any vite-based Storybook, set up the addon via the `npx storybook@next add @storybook/experimental-addon-vitest` command
3 - On any vite-based Storybook, set up the addon by following the docs here https://web-git-8-3-pre-release-storybookjs.vercel.app/docs/8.3/writing-tests/test-runner-with-vitest from this PR: https://github.com/storybookjs/storybook/pull/28924

And run vitest to execute the tests. You can also test the [VSCode/Jetbrains IDE integrations](https://vitest.dev/guide/ide).

#### The changes in this PR are covered in the following automated tests:
- [x] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This pull request has been released as version `0.0.0-pr-28768-sha-416cd00d`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-28768-sha-416cd00d sandbox` or in an existing project with `npx storybook@0.0.0-pr-28768-sha-416cd00d upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-28768-sha-416cd00d`](https://npmjs.com/package/storybook/v/0.0.0-pr-28768-sha-416cd00d) |
| **Triggered by** | @kasperpeulen |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`vitest-integration`](https://github.com/storybookjs/storybook/tree/vitest-integration) |
| **Commit** | [`416cd00d`](https://github.com/storybookjs/storybook/commit/416cd00dcb522130597d92e30e130a46217bff6d) |
| **Datetime** | Wed Aug 14 14:11:53 UTC 2024 (`1723644713`) |
| **Workflow run** | [10389237181](https://github.com/storybookjs/storybook/actions/runs/10389237181) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=28768`_
</details>

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  76.3 MB | 76.3 MB | 0 B | **1.36** | 0% |
| initSize |  167 MB | 169 MB | 1.43 MB | **62.64** | 0.8% |
| diffSize |  91.1 MB | 92.6 MB | 1.43 MB | **178.06** | 1.5% |
| buildSize |  7.42 MB | 7.46 MB | 41.7 kB | **12961.41** | 0.6% |
| buildSbAddonsSize |  1.61 MB | 1.61 MB | -8 B | **-2.46** | 0% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  2.29 MB | 2.3 MB | 14.6 kB | **1104.9** | 0.6% |
| buildSbPreviewSize |  351 kB | 350 kB | -506 B | **-177.56** | -0.1% |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  4.45 MB | 4.46 MB | 14.1 kB | **2190.77** | 0.3% |
| buildPreviewSize |  2.97 MB | 3 MB | 27.6 kB | **8579.87** | 0.9% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  25.8s | 7s | -18s -810ms | -1.16 | -266.9% |
| generateTime |  20.8s | 19.5s | -1s -293ms | -0.14 | -6.6% |
| initTime |  19s | 16.6s | -2s -333ms | -0.23 | -14% |
| buildTime |  10.6s | 12.2s | 1.5s | 0.36 | 13% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  7.2s | 8.5s | 1.2s | 0.26 | 14.9% |
| devManagerResponsive |  4.7s | 5.1s | 434ms | 0.18 | 8.4% |
| devManagerHeaderVisible |  841ms | 759ms | -82ms | -0.76 | -10.8% |
| devManagerIndexVisible |  885ms | 794ms | -91ms | -0.7 | -11.5% |
| devStoryVisibleUncached |  1.3s | 1.3s | -54ms | 0.09 | -4% |
| devStoryVisible |  886ms | 807ms | -79ms | -0.76 | -9.8% |
| devAutodocsVisible |  773ms | 715ms | -58ms | -0.47 | -8.1% |
| devMDXVisible |  658ms | 709ms | 51ms | -0.42 | 7.2% |
| buildManagerHeaderVisible |  705ms | 735ms | 30ms | -0.19 | 4.1% |
| buildManagerIndexVisible |  713ms | 742ms | 29ms | -0.16 | 3.9% |
| buildStoryVisible |  755ms | 778ms | 23ms | -0.22 | 3% |
| buildAutodocsVisible |  607ms | 619ms | 12ms | -0.97 | 1.9% |
| buildMDXVisible |  576ms | 638ms | 62ms | -0.43 | 9.7% |

## Greptile Summary

The pull request integrates Vitest into the Storybook codebase, introducing new events and components for enhanced filtering capabilities.

- **MIGRATION.md**: Updated to document changes from 8.2.x to 8.3.x, including the removal and deprecation of certain addon types.
- **code/core/src/core-events/index.ts**: Added `SET_FILTER` event to the events enum.
- **code/core/src/manager-api/modules/stories.ts**: Introduced `SET_FILTER` event for dynamic story filtering.
- **code/core/src/manager/components/sidebar/FilterToggle.tsx**: Added a new `FilterToggle` component for the sidebar.
- **code/core/src/manager/components/sidebar/Sidebar.tsx**: Removed `bottom` prop, simplifying the component structure.

### CODE REVIEW

File: docs/api/portable-stories/portable-stories-vitest.mdx [hunk 1]
```diff
@@ -170,25 +170,18 @@ sidebar:
...
 
   {/* prettier-ignore-end */}
 
-  <Callout variant="warning">
-
-  You **must** provide `testingLibraryRenderer` to use portable stories with Vitest.
-
-  </Callout>
-
   {/* TODO: Create issue for interest in non-Testing Library render option, with recipe, and mention here (Jest, too) */}
```
Comment (kylegach) [inline docs/api/portable-stories/portable-stories-vitest.mdx:185]:
```suggestion
```

Comment (kylegach) [reply docs/api/portable-stories/portable-stories-vitest.mdx:185]: We should also consider closing the [issue](https://github.com/storybookjs/storybook/discussions/28532)?

It's had zero engagement.

**Comment by radeva** [comment]
Hi, I was testing our dev tool, FlyCI Wingman, with your failing PR, and it managed to provide a fix for your failing tests. Here is the [FlyCI Wingman report](https://github.com/radeva/storybook/pull/1#issuecomment-2273630996) we generated. If you find it useful, feel free to use the provided patch. You can also easily integrate FlyCI Wingman into your repository and get recommendations for fixes by yourself. Documentation can be found at: https://flyci.net/docs.

File: code/addons/vitest/src/plugin/setup-file.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  if (
+    currentTask.type === 'test' &&
+    currentTask.result?.state === 'fail' &&
+    meta.storyId &&
+    currentTask.result.errors?.[0]
+  ) {
+    const currentError = currentTask.result.errors[0];
+    const storybookUrl = import.meta.env.__STORYBOOK_URL__;
+    const storyUrl = `${storybookUrl}/?path=/story/${meta.storyId}&addonPanel=storybook/interactions/panel`;
+    // TODO: figure out why this is being called twice. For now we just do not modify the message again
```
Comment (kasperpeulen) [inline code/addons/vitest/src/plugin/setup-file.ts:32]: Can be removed now right?

File: code/addons/vitest/src/plugin/test-utils.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+/* eslint-disable no-underscore-dangle */
+import type { RunnerTask, TaskContext, TaskMeta } from 'vitest';
+
+import type { ComposedStoryFn } from 'storybook/internal/types';
+
+import type { UserOptions } from './types';
+import { setViewport } from './viewports';
+
+export { setViewport } from './viewports';
```
Comment (kasperpeulen) [inline code/addons/vitest/src/plugin/test-utils.ts:11]: Not sure if this is needed

File: code/addons/vitest/src/plugin/setup-file.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+const modifyErrorMessage = (currentTask: RunnerTask) => {
+  const meta = currentTask.meta as ExtendedMeta;
+  if (
+    currentTask.type === 'test' &&
+    currentTask.result?.state === 'fail' &&
+    meta.storyId &&
+    currentTask.result.errors?.[0]
+  ) {
+    const currentError = currentTask.result.errors[0];
+    const storybookUrl = import.meta.env.__STORYBOOK_URL__;
```
Comment (kasperpeulen) [inline code/addons/vitest/src/plugin/setup-file.ts:30]: Is this correct, when we are starting storybook ourself?

**Review by kasperpeulen — COMMENTED** [review]
Some comments

File: code/addons/vitest/src/plugin/types.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+   *  Whether to skip running Storybook when running tests in watch mode.
+   *  Only relevant when `storybookScript` is provided.
+   *  @default false
+   */
+  skipRunningStorybook?: boolean;
+  /**
+   *  Whether to generate DOM snapshots for every story file.
+   *  @default false
+   */
+  snapshot?: boolean;
```
Comment (kasperpeulen) [inline code/addons/vitest/src/plugin/types.ts:24]: This one doesn't make sense anymore

File: code/addons/vitest/src/plugin/setup-file.ts (hunk 2 above)
Comment (yannbf) [reply code/addons/vitest/src/plugin/setup-file.ts:32]: yes

File: code/addons/vitest/src/plugin/viewports.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  // eslint-disable-next-line no-var, @typescript-eslint/naming-convention
+  var __vitest_browser__: boolean;
+}
+
+interface ViewportsParam {
+  defaultViewport: string;
+  viewports: ViewportMap;
+}
+
+export const setViewport = async (viewportsParam: ViewportsParam = {} as ViewportsParam) => {
```
[the remark this answers is not in the document]
Comment (yannbf) [reply code/addons/vitest/src/plugin/viewports.ts:17]: Let's experiment later.

**Review by yannbf — APPROVED** [review]

**Pull request merged**